src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -848,29 +842,39 @@ MulticopterPositionControl::reset_alt_sp()
 void
 MulticopterPositionControl::limit_altitude()
 {
-	if (_vehicle_land_detected.alt_max < 0.0f) {
-		// there is no altitude limitation present
-		return;
-	}
 
-	float altitude_above_home = -(_pos(2) - _home_pos.z);
+	// Calculate altitude limit
+	float altitude_limit = INFINITY;
+
+	if (_terrain_follow) {
+		altitude_limit = fmaxf(-_local_pos.hagl_max, -_vehicle_land_detected.alt_max);
+
+		//printf("terrain follow alt limit : %f\n", altitude_limit);
+
+	} else {
+		altitude_limit = fmaxf(-_local_pos.hagl_max + _pos(2) + _local_pos.dist_bottom,
+				       -_vehicle_land_detected.alt_max + _home_pos.z);
 
-	if (_run_alt_control && (altitude_above_home > _vehicle_land_detected.alt_max)) {
-		// we are above maximum altitude
-		_pos_sp(2) = -_vehicle_land_detected.alt_max +  _home_pos.z;
+		//printf("altitude follow alt limit : %f\n", altitude_limit);
+	}
 
-	} else if (!_run_alt_control && _vel_sp(2) <= 0.0f) {
-		// we want to fly upwards: check if vehicle does not exceed altitude
+	// Don't allow the height setpoint to exceed the limits
+	if (_pos_sp(2) < altitude_limit) {
+		_pos_sp(2) = altitude_limit;
+	}
+
+	// We want to fly upwards. Check that vehicle does not exceed altitude
+	if (!_run_alt_control && _vel_sp(2) <= 0.0f) {
```

Comment:
This code assumes a constant accel braking manoeuvre which is an invalid assumption for a linear controller.

Reply:
@Stifael @MaEtUgR